primary/src/tests/core_tests.rs

```diff
@@ -126,6 +127,25 @@ async fn process_header() {
         metrics.headers_processed.get_metric_with(&m).unwrap().get(),
         1
     );
+
+    // Test idempotence by re-sending the same header and expecting the vote
+
+    // Send the header to the core again.
+    tx_primary_messages
+        .send(PrimaryMessage::Header(header.clone()))
+        .await
+        .unwrap();
+
+    // Ensure the listener correctly received the vote again.
+    match handle.recv().await.unwrap() {
+        PrimaryMessage::Vote(x) => assert_eq!(x, expected),
+        x => panic!("Unexpected message: {:?}", x),
+    }
+
+    assert_eq!(
+        metrics.headers_processed.get_metric_with(&m).unwrap().get(),
+        2
+    );
```

Comment:
Great having this test here! I think it would be great to add another test that showcases the behaviour when a header of:
* less round is sent
* same round but leading to a different vote digest from the one stored

Reply:
I did test the same round leading to a different vote digest case, and noted that when that happens we return out of the code path before reaching this `process_header` function we've updated here. The checks about the digest matching the previous serve as a backup check. I couldn't find a good way in a unit test to verify that nothing happened, which is what we want it to do and what I observed it does. If we at least got to this function I would use the metric, but since it exits before the metric also can't be used to verify following the regular code flow.

Reply:
To be fair, what's meant to happen when using a different round or a different digest is nothing (i.e. we let the vote proceed), something which we do test in many other places.